### ISSUE TITLE
Reactive forms cleared input becomes empty string instead of null 

### ISSUE THREAD

**Issue by aleripe** [angular/angular#45317]
### Which @angular/* package(s) are the source of the bug?

forms

### Is this a regression?

No

### Description

I have created a minimal reproducible sample of the problem here: https://stackblitz.com/edit/angular-ivy-4ltwow

Here you can see the mocked model input (that I receive from server API in the real app) is null at first, but after I edit the value in textbox and clear it, it becomes empty string.

I test this value server side to see if it has changed since before, and empty string is different from null so I have to test the difference from null AND empty string literally everywhere.

I have no alternative other than creating a directive and applying on every text input, but there are literally thousands and if I forget to put only one the problem arises again.

Shouldn't `getRawValue` return null if the input control is blanked? Wouldn't it be possibile to include an option in the form initialization (or a parameter of `getRawValue`) to get this desired behavior?

Thanks! 

### Please provide a link to a minimal reproduction of the bug

https://stackblitz.com/edit/angular-ivy-4ltwow

### Please provide the environment you discovered this bug in (run `ng version`)

```true
Angular CLI: 9.0.2 (but I tested up to Angular 13 and it's the same)
Node: 16.14.0
OS: win32 x64

Angular: 9.0.1
... animations, common, compiler, compiler-cli, core, forms       
... language-service, platform-browser, platform-browser-dynamic  
... router
Ivy Workspace: Yes
```

**Comment by JoostK** [comment]
> Here you can see the mocked model input (that I receive from server API in the real app) is null at first, but after I edit the value in textbox and clear it, it becomes empty string.

What do you mean by "clear it"? Just removing the data with backspace? Then it makes sense to me that the resulting data stays the empty string.

> Shouldn't getRawValue return null if the input control is blanked? Wouldn't it be possibile to include an option in the form initialization (or a parameter of getRawValue) to get this desired behavior?
This is highly dependable on how projects deal with data; a `null` value isn't necessarily more desirable than an empty string.

An option might work but seems out-of-place to me. This use-case would likely fit under #3009, which can be achieved today using a custom value accessor. You should be able to write a directive similar to [`DefaultValueAccessor`](https://github.com/angular/angular/blob/a08ea3f6433b842b0e342509e5ddd53907a863b0/packages/forms/src/directives/default_value_accessor.ts) that provides [`NG_VALUE_ACCESSOR`](https://angular.io/api/forms/NG_VALUE_ACCESSOR) to override the default `DefaultValueAccessor`, which can then convert empty strings to null as you desire. This would be more ergnomic with #3009, but that is not yet available.

**Comment by aleripe** [comment]
Yes, I mean using the backspace. It seems to me counter intuitive that upon patching the control value is null, and after having been cleared it becomes an empty string.
I will investigate the DefaultValueAccessor directive. #3009 has been opened since 2015 so I don't think it will happen soon...

Thanks!

**Comment by aleripe** [comment]
I see this method in `DefaultValueAccessor`: 

```
  writeValue(value: any): void {
    const normalizedValue = value == null ? '' : value;
    this.setProperty('value', normalizedValue);
  }
```

So shouldn't the value be '' (empty string) since the first patching? I find this behavior inconsistent...

**Comment by JoostK** [comment]
Yeah, writing the `value` attribute doesn't "bounce back" to the control value.

**Comment by yelhouti** [comment]
@kara sorry for the ping but since you are the one who wrote this (6 years ago :) ), I was wondering if you remember the reason why you set null(ish) values to the empty string:
https://github.com/angular/angular/blob/a08ea3f6433b842b0e342509e5ddd53907a863b0/packages/forms/src/directives/default_value_accessor.ts#L103-L106

Thank you for your help

**Comment by yelhouti** [comment]
The issue seems to come directly from how the `oninput` event is handled in html/javascript. the `event.target.value` of an empty input field is `''` (an empty string).
I understand why one would like to change this behavior for all its fields (I a came here for that too) and although I think it's a good default I understand that is (kind of) deviating from the vanilla javascript can be scary and is huge breaking change.

In the meantime, you can just create this directive in some "CommonModule" and import it in all your modules and it should work.
```
import { Directive, ElementRef, forwardRef, HostListener, Renderer2 } from '@angular/core';
import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';

export const EMPTY_TO_NULL_VALUE_ACCESSOR: any = {
  provide: NG_VALUE_ACCESSOR,
  useExisting: forwardRef(() => EmptyToNullValueAccessor),
  multi: true
};

@Directive({
  // eslint-disable-next-line @angular-eslint/directive-selector
  selector: 'input:not([type=checkbox])[formControlName],textarea[formControlName],input:not([type=checkbox])[formControl],textarea[formControl],input:not([type=checkbox])[ngModel],textarea[ngModel],[ngDefaultControl]',
  providers: [EMPTY_TO_NULL_VALUE_ACCESSOR]
})
// eslint-disable-next-line @angular-eslint/directive-class-suffix
export class EmptyToNullValueAccessor implements ControlValueAccessor {

  constructor(private renderer: Renderer2, private elementRef: ElementRef) {
  }

  // eslint-disable-next-line @typescript-eslint/no-empty-function
  onChange = (_: any): void => {};
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  onTouched = (): void => {};

  registerOnChange(fn: (_: string|null) => void): void {
    this.onChange = (value) => { fn(value === '' ? null : value); };
  }

  registerOnTouched(fn: any): void {
    this.onTouched = fn;
  }

  setDisabledState(isDisabled: boolean): void {
    this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
  }

  writeValue(value: any): void {
    const normalizedValue = value == null ? '' : value;
    this.renderer.setProperty(this.elementRef.nativeElement, 'value', normalizedValue);
  }

  @HostListener('change', ['$event']) change(event: any): void {
    this.onChange(event.target.value);
  }

  @HostListener('input', ['$event']) input(event: any): void {
    this.onChange(event.target.value);
  }

  @HostListener('blur') blur(): void {
    this.onTouched();
  }
}

```

**Comment by yelhouti** [comment]
Also I am willing to make a PR for this if the core team thinks it should be changed for the next major version

**Comment by jarodsmk** [comment]
Glad I came across this - 

In my case, I read the information from my strongly-typed `FormGroup` with `.getRawValues()` and my `<number | null>FormControl` instances now have empty string values after being cleared as discussed above. The result of this is down the line, POST-ing my request to my API  results in an `application/problem+json` media type, which my API promptly rejects (being unsupported).

We made a design decision to not use `type="number"` inputs, as we want a user to focus on a variety of inputs and actually type values in, thus making a 0 value explicit from their side. Worth mentioning this isn't a public-facing system, and our inputs aren't custom components.

**Comment by sebastianhaberey** [comment]
@jarodsmk this is exactly what we're experiencing: strongly typed `FormControl` created as `this.fb.control<string | undefined>(undefined)`. Expecting default value (undefined) when cleared by user. Getting empty string.

**Comment by i-am-the-slime** [comment]
You say strongly typed but you don't seem to be using: https://gcanti.github.io/io-ts-types/modules/NonEmptyString.ts.html
